cpp/src/gandiva/hash_utils.cc

```diff
@@ -0,0 +1,137 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstring>
+#include "openssl/evp.h"
+#include "gandiva/hash_utils.h"
+#include "gandiva/execution_context.h"
+#include "gandiva/gdv_function_stubs.h"
+
+namespace gandiva {
+  const char* HashUtils::HashUsingSha256(int64_t context,
+                                         const void* message,
+                                         size_t message_length,
+                                         int32_t *out_length) {
+    // The buffer size is the hash size + null character
+    int sha256_result_length = 65;
+    return HashUtils::GetHash(context, message, message_length, EVP_sha256(),
+							  sha256_result_length, out_length);
+  }
+  const char* HashUtils::HashUsingSha128(int64_t context,
+                                         const void* message,
+                                         size_t message_length,
+                                         int32_t *out_length) {
+    // The buffer size is the hash size + null character
+    int sha128_result_length = 41;
+    return HashUtils::GetHash(context, message, message_length, EVP_sha1(),
+							  sha128_result_length, out_length);
+  }
+
+  const char* HashUtils::GetHash(int64_t context,
+                                 const void* message,
+                                 size_t message_length,
+                                 const EVP_MD *hash_type,
+                                 int result_buf_size,
+                                 int32_t *out_length) {
+    EVP_MD_CTX *md_ctx = EVP_MD_CTX_new();
+
+    if (md_ctx == nullptr) {
+      HashUtils::ErrorMessage(context, "Could not allocate memory "
+									   "for SHA processing.");
+      return "";
+    }
+
+    int evp_success_status = 1;
+
+    if (EVP_DigestInit_ex(md_ctx, hash_type, nullptr) != evp_success_status) {
+      HashUtils::ErrorMessage(context, "Could not obtain the hash "
+									   "for the defined value.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    if (EVP_DigestUpdate(md_ctx, message, message_length) != evp_success_status) {
+      HashUtils::ErrorMessage(context, "Could not obtain the hash for "
+									   "the defined value.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    int hash_size = EVP_MD_size(hash_type);
+    auto* result = static_cast<unsigned char*>(OPENSSL_malloc(hash_size));
+
+    if (result == nullptr) {
+      HashUtils::ErrorMessage(context, "Could not allocate memory "
+									   "for SHA processing.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    unsigned int result_length;
+    EVP_DigestFinal_ex(md_ctx, result, &result_length);
```

Comment:
add a check result_length == hash_size
and result_buf_size == 2 * hash_size

Reply:
Checks added.

Reply:
Checks added